File: Competitive Programming 3/Exercises/Introduction/Getting Started (The Easy Problems)/Medium/Greedy Gift Givers/3g.cpp
#include <iostream>
#include <map>
#include <vector>
#include <string>
#include <sstream>
using namespace std;

int main() {
    const string given = "given";
    const string taken = "taken";
    int motherF = 0;

    while (true) {
        int n;
        if (!(cin >> n)) {
            break;
        }
        if (motherF != 0) {
            cout << endl;
        }
        cin.ignore();

        map<string, map<string, int>> friends;
        string line;
        getline(cin, line);
        istringstream namesStream(line);
        vector<string> names;
        string name;

        while (namesStream >> name) {
            names.push_back(name);
            friends[name][given] = 0;
            friends[name][taken] = 0;
        }

        for (int i = 0; i < n; ++i) {
            getline(cin, line);
            istringstream queryStream(line);
            string giver;
            int amount, num;
            queryStream >> giver >> amount >> num;

            int mon = (num != 0) ? amount / num : 0;
            friends[giver][given] = (mon * num);

            for (int j = 0; j < num; ++j) {
                string receiver;
                queryStream >> receiver;
                friends[receiver][taken] += mon;
            }
        }

        for (const string& name : names) {
            cout << name << " " << (friends[name][taken] - friends[name][given]) << endl;
        }
        motherF += 1;
    }
    return 0;
}

File: Competitive Programming 3/Exercises/Introduction/Getting Started (The Easy Problems)/Medium/Greedy Gift Givers/3g.py
given = 'given'
taken = 'taken'
mf = 0
while True:
    try:
        n = int(input())
        if mf != 0:
            print()
        friends = dict()
        names = input().split(' ')
        for i in range(n):
            friends[names[i]] = dict()
            friends[names[i]][given] = 0
            friends[names[i]][taken] = 0
        for _ in range(n):
            query = input().split(' ')
            num = int(query[2])
            mon = (int(query[1]) // num) if num != 0 else 0
            friends[query[0]][given] = (mon * num)
            while num > 0:
                friends[query[num + 2]][taken] += mon
                num -= 1
        for i in names:
            print(f'{i} {friends[i][taken] - friends[i][given]}')
        mf += 1
    except EOFError:
        break